src/helpers/get-items-from-portal.js

```diff
@@ -1,46 +1,56 @@
 const axios = require('axios');
 const { serializeQueryParams } = require('./portal-query-builder');
 
-async function getPortalItems(portalUrl, query, maxPageSize) {
-    const firstPage = await fetchItemsFromPortal(portalUrl, query);
-    const totalBatch = getTotalBatch(firstPage.total, maxPageSize);
+async function getPortalItems(portalItemsRequestOptions, logOptions) {
+    const { portalUrl, portalQuery, MAX_PAGE_SIZE } = portalItemsRequestOptions;
+    const firstPage = await fetchItemsFromPortal(portalUrl, portalQuery, logOptions);
+    const totalBatch = getTotalBatch(firstPage.total, MAX_PAGE_SIZE);
     if (totalBatch === 1) {
         return {
             items: firstPage.results,
             count: firstPage.total
         };
     }
-    const remainingRequests = buildRemainingPageRequests({ portalUrl, query, totalBatch, maxPageSize });
-    const remainingItems = await getRemainingPortalItems(remainingRequests);
+    const remainingRequests = buildRemainingPageRequests({ portalUrl, portalQuery, totalBatch, MAX_PAGE_SIZE });
+    const remainingItems = await getRemainingPortalItems(remainingRequests, logOptions);
     return {
         items: [...firstPage.results, ...remainingItems],
         count: firstPage.total
     };
 };
 
-async function fetchItemsFromPortal(portalUrl, query) {
-    const url = `${portalUrl}?${serializeQueryParams(query)}`;
+async function fetchItemsFromPortal(portalUrl, portalQuery, logOptions) {
+    const { log, logLevel } = logOptions;
+    const url = `${portalUrl}?${serializeQueryParams(portalQuery)}`;
     const items = await axios.get(url);
+    if (logLevel) {
+        log[logLevel](`Request made to ${url}`);
+    }
     return items.data;
 }
 
 function getTotalBatch(total, maxPageSize) {
     return Math.ceil(total / maxPageSize);
 }
 
-function buildRemainingPageRequests({ portalUrl, query, totalBatch, maxPageSize }) {
+function buildRemainingPageRequests(requestOptions) {
+    const { portalUrl, portalQuery, totalBatch, MAX_PAGE_SIZE } = requestOptions;
     const requests = [];
     // Multiple batches each with maxPageSize number of records
     for (let i = 1; i < totalBatch; i++) {
-        query.start += maxPageSize;
-        requests.push(axios.get(`${portalUrl}?${serializeQueryParams(query)}`));
+        portalQuery.start += MAX_PAGE_SIZE;
+        requests.push(axios.get(`${portalUrl}?${serializeQueryParams(portalQuery)}`));
     }
     return requests;
 }
 
-async function getRemainingPortalItems(requests) {
+async function getRemainingPortalItems(requests, logOptions) {
+    const { log, logLevel } = logOptions;
     const pages = await Promise.all(requests);
     const items = pages.reduce((collection, page) => {
+        if (logLevel) {
+            log[logLevel](`Request made to https://${page.request.host}${page.request.path}`);
```

Comment:
Here you have the log _after_ the promise.all.  I think that means if any single request fails, you won't get your logs.  The logging has to happen _before_ the request is made to Portal.

You can wrap the `axios.get(`${portalUrl}?${serializeQueryParams(portalQuery)}` in another async function that include the log prior to the `axios.get`.

Reply:
Got it! It makes sense to log after the request. I have made the required changes. I used log before creating axios request promises.

Reply:
@rgwozdz : I have included wrapped `axios.get` in another async function and included log there as suggested.